### ISSUE TITLE
iOS Browser

### ISSUE THREAD

**Issue by nbarkhina** [aspnet/Blazor#73]
Steve mentioned that mobile browsers are supported however none of the demos load on iOS 11 - for me they only work in Edge and Chrome on Desktop. Is iOS currently working or is that a future feature?

**Comment by dantheman999301** [comment]
I saw Steve demo it on IE11, so I'm surprised it doesn't work on iOS 11.

**Comment by nbarkhina** [comment]
I saw it too and can confirm it doesn't work for me in IE11 either. His demo was from last summer though so maybe something changed in the code.

**Comment by StephanHartmann** [comment]
I can confirm that the sample link doesn't work on my iPad too (current iOS)

**Comment by SteveSandersonMS** [comment]
I've verified the http://blazor-demo.azurewebsites.net/ site works correctly in both IE11 and in Safari on iOS 11, so I'm not sure what issues you're hitting. If you manage to debug and track down a specific error please let us know.

**Comment by StephanHartmann** [comment]
I tested it right now.
It got stuck displaying "Loading...".
My device is no developper device, so I don't know how to debug on it, sorry.

**Comment by natemcmaster** [comment]
This is what I'm seeing

![image](https://user-images.githubusercontent.com/2696087/35985991-c49c3df0-0cac-11e8-9799-3eeb924dce53.png)

**Comment by SteveSandersonMS** [comment]
Oh, I see what's going on. Although it works fine in iOS 11.1, it fails in 11.2 due to this: https://github.com/kripken/emscripten/issues/6042

TLDR: Apple regressed WebAssembly in iOS 11.2 due to Spectre/Meltdown mitigation. It affects all Emscripten builds.

I don't think it's relevant to track this in this repo since it's a huge external issue affecting all WebAssembly apps that inevitably will be fixed externally, so I'll close this. In the very unlikely event that it doesn't get fixed by the time we publish a download, we'll certainly know about it and can make it fall back on asm.js for iOS.

**Comment by Daddoon** [comment]
Still throwing on iOS 11.3 (just tested).

**Comment by Daddoon** [comment]
Thanks for the info Steve.
I'm actually trying to manage to host a Blazor app directly on a mobile with Xamarin as interoperability.

At first i was not sure if it was about me or a bug. Seem it's Safari fault.

**Comment by GoranHalvarsson** [comment]
Any news on this? This is important, there are a lot of iPhone users out there.

**Comment by Daddoon** [comment]
Actually, on Safari (iOS) you have 2 kinds of problem:

- Using UIWebview component, WebAssembly is still buggy, the bug can be detected with a minimal test case and so you have to fallback to asmjs in this case.
- Using WkWebview component does crash like 2 seconds after having loaded Blazor, like an unexpected close, without much info. Fallback to asmjs correct the issue, but i don't have any test case to detect if the bug is present.

I don't know wich one the "Safari Browser" implementation is used directly from the Safari app.

You may take a try to my [Blazor.Polyfill](https://github.com/Daddoon/Blazor.Polyfill) project wich take care of theses usecases.

Tell me if it go right for you!

**Comment by Andrzej-W** [comment]
It looks they have found the bug today.
https://bugs.webkit.org/show_bug.cgi?id=184829
Now we have to wait for official update.

**Comment by Daddoon** [comment]
@thesoftwarejedi Can you please post the useragent string of your Chrome version on iOS ?

**Comment by chucker** [comment]
>Using UIWebview component, WebAssembly is still buggy, the bug can be detected with a minimal test case and so you have to fallback to asmjs in this case.

UIWebView is [deprecated](https://twitter.com/bradeeoh/status/1009302764722184192) now, FWIW.

>I don't know wich one the "Safari Browser" implementation is used directly from the Safari app.

Safari is closer to `WKWebView` (but possibly internally its own thing). WKWebView and Safari are both WebKit2, i.e. a multi-process architecture, whereas UIWebView was single-process.

**Comment by Daddoon** [comment]
Thanks for the feedback @chucker !

On my **Blazor.Xamarin** project, i have ended to use WkWebview at the end. And for Web browsing, we can assume that newer version of Safari always use the WkWebview architecture.

Still keeping the WebAssembly bug fix on UIWebview (for related buggy version) in the **Blazor.Polyfill** project as maybe we can encounter someone using some old version if iOS cannot be updated on his device (also, i'm not following Apple news, so i have no clue which phone is still updated or not.

Still good to know theses informations, thanks a lot!

**Comment by chucker** [comment]
>iOS 11.4 here, this is still an issue. Blazor does not work on iPhone Chrome.

Blazor will not currently work on iOS (including the current betas of 11.4.x and 12), no matter the browser, due to the aforementioned bug in WebKit.

>Still keeping the WebAssembly bug fix on UIWebview (for related buggy version) in the Blazor.Polyfill project as maybe we can encounter someone using some old version if iOS cannot be updated on his device (also, i'm not following Apple news, so i have no clue which phone is still updated or not.

Right. Since the bug is fixed in trunk, the final version of iOS 12 will _presumably_ have the fix, so you should probably include a version check this fall.

**Comment by Daddoon** [comment]
It would be nice to know the minimal Webkit buggy version, and the actual (at least). As you said a version check will surely be necessary this fall, as we cannot know the exact minimal non-buggy version number when all theses thing will be released.

I have to do some check...We can assume that:

- 605.1.15 : **KO**
- 605.1.33 : **KO**
- 604.1.38 : **Unknown** (Maybe ok on the current bug, but has maybe the other WebAssembly bug, managed in my project
- 604.5.6 : **OK**  , as stated by [Rodrigo Kumpera](https://bugs.webkit.org/show_bug.cgi?id=184829)

So at the moment, i think we have to fallback to asmjs on versions between 605.1.x to 605.1.33 maximum. Not that i check the AppleWebkit version number as someone stated the bug exist on Chrome too on the related Webkit versions.

I will do some check.

**Comment by Daddoon** [comment]
Just shipped the workaround to fallback on **asmjs** on buggy Webkit versions. It's of course not ideal for the performance drawback, but i does not crash anymore on theses specific version on **Chrome** and **Safari**.

If needed, take a look here: [Blazor.Polyfill](https://github.com/Appizeo/Blazor.Polyfill)
